Whimsy rework

spwilson2 opened the pull request:
Like I said... Huge changes... Sorry about that. 

Luckily, gem5_verify_config doesn't change at all to test writers. (Other than the fact that it works.) Feel free to tell me if the changes are cruft (Or if any portion of them are.) Besides everything actually working, in my experience this implementation has been a lot easier to debug.

I've really only done preliminary integration testing. All the tests are passing so that's pretty neat...

A good quick command I've used to test execution is `./main.py run gem5/ -vvv --include-tags opt` 

If you'd like me to go through any of my thought process with these changes or a provide an overview via videochat let me know. My schedule this coming weekend will be full, but the middle of the week I should be available.

Whimsy 1.0 known issues this rework fixes:
- JUnit XML output is working now
- SCons Build correctly creates minimal set of targets
- Verifiers are correctly instantiated for `gem5_verify_config` suites
- Fixture setup failures result in library/suite/test error, and display message.


## TODOs


So there are a couple things I already know I still have to get working:
- ~~Rerun command~~
- ~~Your newbie friendly `isa`, `variant`, and `length` tags.~~
- ~~Multi-threaded test suite execution~~
- JUnit XML error, skip results don't supply the fixture setup exception that caused the result.
- I'll need to look at merging in my unit test patch, I don't think that will be a big change, just a refactor.
- ~~Adding license headers to the source.~~

If I have time Tuesday night I'll also be sure to start re-writing documentation. 

Finally, if I get all the above sorted out I'd like to take a stab at refactoring that config. I'd like to get all gem5 specific configuration out of the test library if possible. It would make it much easier for me to continue to work on library improvements and import/export patches in the future.

spwilson2 commented:
Restored:
- rerun
- `isa`, `variant`, and `length` tags
- multithreaded/parallel suite execution
- license
- some documentation (still needs more)


---
Only bug I've found so far is that single fixture instances attached to multiple tests/suites will be built multiple times. This could cause issues in parallel test execution, as a solution I intend to require fixtures implement a copy method which will be executed when they receive the test schedule. This way setup/teardown will be performed on a separate instance each time. 

I should have a fix for this in tonight.

powerjg commented:
I haven't had a chance to go through this yet, but I have a quick question... 

Should I hold off on pushing what we have now on gerrit for these changes?

powerjg commented:
I hope to be able to go through this this afternoon.

spwilson2 commented:
Yeah I would hold off. I think these changes are about ready to merge into
our Gerrit patchset.

As it is now, the fixture logic and verifier fixes make the library much
more reliable to use and program against.

On Thu, Jul 12, 2018, 13:22 Jason Lowe-Power <notifications@github.com>
wrote:

> I hope to be able to go through this this afternoon.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jlpresearch/gem5/pull/5#issuecomment-404605025>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AK9rOMB7jUS7B6oXSDqQg9MflGJv7-xAks5uF5PXgaJpZM4VG932>
> .
>

powerjg commented:
Sounds good. I'll try to take a deeper look this afternoon.

FYI, a quick glance shows some style issues (e.g., lines with more than 79
chars and newlines missing at the end of files).

Cheers,
Jason

On Thu, Jul 12, 2018 at 11:29 AM Sean Wilson <notifications@github.com>
wrote:

> Yeah I would hold off. I think these changes are about ready to merge into
> our Gerrit patchset.
>
> As it is now, the fixture logic and verifier fixes make the library much
> more reliable to use and program against.
>
> On Thu, Jul 12, 2018, 13:22 Jason Lowe-Power <notifications@github.com>
> wrote:
>
> > I hope to be able to go through this this afternoon.
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/jlpresearch/gem5/pull/5#issuecomment-404605025>, or
> mute
> > the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AK9rOMB7jUS7B6oXSDqQg9MflGJv7-xAks5uF5PXgaJpZM4VG932
> >
> > .
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/jlpresearch/gem5/pull/5#issuecomment-404607207>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGsHCfQ077yrEsgoKeqbDW1M39csqPuks5uF5WUgaJpZM4VG932>
> .
>


-- 
Jason Lowe-Power
Assistant Professor, Computer Science Department
University of California, Davis
3049 Kemper Hall
https://faculty.engineering.ucdavis.edu/lowepower/

spwilson2 commented:
Oops, yeah forgot to re-enable style-check. I'' be sure to give it a run through.

Also, the last thing I'd really like to get back in is the `--uid` flag for the run command. After that's back I don't have any blocking issues with this being merged.

powerjg commented:
Hey Sean,

Sorry this took so long. I've finally had some time to test it out.

A few things:
1) Could you rebase this branch so it can be pushed to mainline? If we want to push with the new whimsy that commit should be amended.
2) There's a bug in python threading that requires a change in tests/gem5/fixtures.py (sorry for the bad formatting, but it'll be easier for you to fix since you're rebasing anyway).
`diff --git a/tests/gem5/fixture.py b/tests/gem5/fixture.py                                                                                                                                                                                      index 084ecc2..4d9b455 100644                                                                                                                                                                                                                   --- a/tests/gem5/fixture.py                                                                                                                                                                                                                     +++ b/tests/gem5/fixture.py                                                                                                                                                                                                                     @@ -219,6 +219,7 @@ class DownloadedProgram(Fixture):                                                                                                                                                                                                                                                                                                                                                                                                                                                def _getremotetime(self):                                                                                                                                                                                                                           import urllib2, datetime, time                                                                                                                                                                                                         +        import _strptime # Needed for python threading bug                                                                                                                                                                                              u = urllib2.urlopen(self.url)                                                                                                                                                                                                                   return time.mktime(datetime.datetime.strptime( \                                                                                                                                                                                                            u.info().getheaders("Last-Modified")[0],`
3) I can't run any tests for ISAs other than NULL, X86, and ARM. There should be one RISC-V test to make sure it compiles (see tests/gem5/test_build.py), but it's not showing up no matter what I do.
I tried `./main.py run --isa RISCV`, but it tires to build all ISAs, but it doesn't list any valid targets under "Building the following targets...". It's definitely a bug, but I'm not sure what's going on.
4) Another small thing while you're rebasing... Could you change the print message "Building the following targets. This may take a while." to "Building the following targets. This may take a while. You may want to run with only a single ISA (--isa=<isa>), use --skip-build, or use 'retest'."
4a) BTW, is retest documented anywhere?
5) Have you gone back through TESTING.md to make sure the documentation is still correct?
6) Try running `./main.py run --uid SuiteUID:tests/gem5/m5_util/test_exit.py:m5_exit_test-X86-fast` This definitely doesn't work correctly.
7) I'm not sure if --include-tags works correctly. 
When I run `./main.py run --include-tags fast -vvv` It doesn't run any of the fast tests. I see this:
`Filtering suites with tags as follows:   Include: fast                                                                                                                                                                                                     Include: ^X86$                                                                                                                                                                                                                                  Include: ^SPARC$                                                                                                                                                                                                                                Include: ^ALPHA$                                                                                                                                                                                                                                Include: ^RISCV$                                                                                                                                                                                                                                Include: ^ARM$                                                                                                                                                                                                                                  Include: ^MIPS$                                                                                                                                                                                                                                 Include: ^POWER$                                                                                                                                                                                                                                Include: ^NULL$                                                                                                                                                                                                                                 Include: ^quick$                                                                                                                                                                                                                                Include: ^opt$                                                                                                                                                                                                                                   Remove: ^long$                                                                                                                                                                                                                                  Remove: ^debug$                                                                                                                                                                                                                                 Remove: ^fast$ `

spwilson2 commented:
Hey Jason,

No worries thanks for reviewing!

I'm out for the weekend at a music festival and don't get back till late
monday night so I won't be able to address this until probably
Wednesday/Thursday.

On Thu, Jul 19, 2018, 19:58 Jason Lowe-Power <notifications@github.com>
wrote:

> Hey Sean,
>
> Sorry this took so long. I've finally had some time to test it out.
>
> A few things:
>
>    1. Could you rebase this branch so it can be pushed to mainline? If we
>    want to push with the new whimsy that commit should be amended.
>    2. There's a bug in python threading that requires a change in
>    tests/gem5/fixtures.py (sorry for the bad formatting, but it'll be easier
>    for you to fix since you're rebasing anyway).
>    diff --git a/tests/gem5/fixture.py b/tests/gem5/fixture.py index
>    084ecc2..4d9b455 100644 --- a/tests/gem5/fixture.py +++
>    b/tests/gem5/fixture.py @@ -219,6 +219,7 @@ class
>    DownloadedProgram(Fixture): def _getremotetime(self): import urllib2,
>    datetime, time + import _strptime # Needed for python threading bug u =
>    urllib2.urlopen(self.url) return time.mktime(datetime.datetime.strptime( \
>    u.info().getheaders("Last-Modified")[0],
>    3. I can't run any tests for ISAs other than NULL, X86, and ARM. There
>    should be one RISC-V test to make sure it compiles (see
>    tests/gem5/test_build.py), but it's not showing up no matter what I do.
>    I tried ./main.py run --isa RISCV, but it tires to build all ISAs, but
>    it doesn't list any valid targets under "Building the following
>    targets...". It's definitely a bug, but I'm not sure what's going on.
>    4. Another small thing while you're rebasing... Could you change the
>    print message "Building the following targets. This may take a while." to
>    "Building the following targets. This may take a while. You may want to run
>    with only a single ISA (--isa=), use --skip-build, or use 'retest'."
>    4a) BTW, is retest documented anywhere?
>    5. Have you gone back through TESTING.md to make sure the
>    documentation is still correct?
>    6. Try running ./main.py run --uid
>    SuiteUID:tests/gem5/m5_util/test_exit.py:m5_exit_test-X86-fast This
>    definitely doesn't work correctly.
>    7. I'm not sure if --include-tags works correctly.
>    When I run ./main.py run --include-tags fast -vvv It doesn't run any
>    of the fast tests. I see this:
>    Filtering suites with tags as follows: Include: fast Include: ^X86$
>    Include: ^SPARC$ Include: ^ALPHA$ Include: ^RISCV$ Include: ^ARM$ Include:
>    ^MIPS$ Include: ^POWER$ Include: ^NULL$ Include: ^quick$ Include: ^opt$
>    Remove: ^long$ Remove: ^debug$ Remove: ^fast$
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/darchr/gem5/pull/5#issuecomment-406457461>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AK9rOGlw1pmltSS5jH5f63H09GRDPxrsks5uISs8gaJpZM4VG932>
> .
>

powerjg commented:
No worries. Have fun!

Jason

On Thu, Jul 19, 2018 at 10:14 PM Sean Wilson <notifications@github.com>
wrote:

> Hey Jason,
>
> No worries thanks for reviewing!
>
> I'm out for the weekend at a music festival and don't get back till late
> monday night so I won't be able to address this until probably
> Wednesday/Thursday.
>
> On Thu, Jul 19, 2018, 19:58 Jason Lowe-Power <notifications@github.com>
> wrote:
>
> > Hey Sean,
> >
> > Sorry this took so long. I've finally had some time to test it out.
> >
> > A few things:
> >
> > 1. Could you rebase this branch so it can be pushed to mainline? If we
> > want to push with the new whimsy that commit should be amended.
> > 2. There's a bug in python threading that requires a change in
> > tests/gem5/fixtures.py (sorry for the bad formatting, but it'll be easier
> > for you to fix since you're rebasing anyway).
> > diff --git a/tests/gem5/fixture.py b/tests/gem5/fixture.py index
> > 084ecc2..4d9b455 100644 --- a/tests/gem5/fixture.py +++
> > b/tests/gem5/fixture.py @@ -219,6 +219,7 @@ class
> > DownloadedProgram(Fixture): def _getremotetime(self): import urllib2,
> > datetime, time + import _strptime # Needed for python threading bug u =
> > urllib2.urlopen(self.url) return time.mktime(datetime.datetime.strptime(
> \
> > u.info().getheaders("Last-Modified")[0],
> > 3. I can't run any tests for ISAs other than NULL, X86, and ARM. There
> > should be one RISC-V test to make sure it compiles (see
> > tests/gem5/test_build.py), but it's not showing up no matter what I do.
> > I tried ./main.py run --isa RISCV, but it tires to build all ISAs, but
> > it doesn't list any valid targets under "Building the following
> > targets...". It's definitely a bug, but I'm not sure what's going on.
> > 4. Another small thing while you're rebasing... Could you change the
> > print message "Building the following targets. This may take a while." to
> > "Building the following targets. This may take a while. You may want to
> run
> > with only a single ISA (--isa=), use --skip-build, or use 'retest'."
> > 4a) BTW, is retest documented anywhere?
> > 5. Have you gone back through TESTING.md to make sure the
> > documentation is still correct?
> > 6. Try running ./main.py run --uid
> > SuiteUID:tests/gem5/m5_util/test_exit.py:m5_exit_test-X86-fast This
> > definitely doesn't work correctly.
> > 7. I'm not sure if --include-tags works correctly.
> > When I run ./main.py run --include-tags fast -vvv It doesn't run any
> > of the fast tests. I see this:
> > Filtering suites with tags as follows: Include: fast Include: ^X86$
> > Include: ^SPARC$ Include: ^ALPHA$ Include: ^RISCV$ Include: ^ARM$
> Include:
> > ^MIPS$ Include: ^POWER$ Include: ^NULL$ Include: ^quick$ Include: ^opt$
> > Remove: ^long$ Remove: ^debug$ Remove: ^fast$
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/darchr/gem5/pull/5#issuecomment-406457461>, or mute
> > the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AK9rOGlw1pmltSS5jH5f63H09GRDPxrsks5uISs8gaJpZM4VG932
> >
> > .
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/darchr/gem5/pull/5#issuecomment-406491260>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGsHDbC-9WB70ZLxaXp44ClPQDz85T4ks5uIWdPgaJpZM4VG932>
> .
>


-- 
Jason Lowe-Power
Assistant Professor, Computer Science Department
University of California, Davis
3049 Kemper Hall
https://faculty.engineering.ucdavis.edu/lowepower/

spwilson2 commented:
Alright finally getting to this...

1. In Progress.
2. Done.
3. Fixed. I changed the tag logic back to just a simple iterable of tags. The definition in `tests/gem5/test_build/test_build.py` was using a dictionary, this has been fixed and now adds a RISCV test+suite.
4. Done. I also added a warning message in the event that no SCons target was specified so users don't unknowingly build every gem5 target.
  I also added a section on rerun to TESTING.md
5. I made a pass through and it looks all good to me. I think I only had to adjust the parallel execution section and the list command.
6. Good catch. This is a bug, the tags are currently filtering out suites when the `--uid` flag is provided. Flag filter no longer occurs when the `--uid` flag is provided, and your usage now works as expected.
7. So this usage is an edge case and I'm not sure how we should deal with it.
  The library implementation expects users to use the `--variant` flag for something like this.
  What happens is that all the gem5 "special" flag filters (opt, fast; quick, long; RISCV, ARM; etc.) are applied after the user specific filters. This makes the gem5 options supplied by the command line take priority.

---
So you don't have to dig through a markdown ordered list for actions to take:

If my comment on number 7 doesn't make sense or your usage should be enabled let me know. Otherwise, I'll work on the rebase which should be good to go.
